org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/GrammarElementsInterner.java

```diff
@@ -0,0 +1,100 @@
+/*******************************************************************************
+ * Copyright (c) 2019 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.nodemodel.impl;
+
+import java.util.Map;
+import java.util.Objects;
+
+import org.eclipse.emf.ecore.EObject;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.ObjectArrays;
+
+/**
+ * Interner of arrays of grammar elements stored in the node model.
+ * 
+ * The node model contains nodes derived from {@link AbstractNode}.  An {@link AbstractNode} has an array describing its grammar 
+ * elements.  This array is often identical for many {@link AbstractNode}s.  There can be millions of nodes in a node model
+ * so it makes sense to remove the duplicate arrays.  This {@link GrammarElementsInterner} tries to reuse an interned version of such an array
+ * wherever possible
+ */
+
+public class GrammarElementsInterner {
+
+	private static class InternKey {
+		final EObject grammarElement;
+		final Object grammarElements;
+		
+		int hashCode = -1;
+
+		InternKey(EObject grammarElement, Object grammarElements) {
+			this.grammarElement = grammarElement;
+			this.grammarElements = grammarElements;
+		}
+		
+		EObject[] prefixElement() {
+			if(grammarElements instanceof EObject) {
+				return new EObject[] {grammarElement, (EObject) grammarElements};
+			} else {
+				return ObjectArrays.concat(grammarElement, (EObject[]) grammarElements);
+			}
+		}
+		
+		EObject[] postfixElement() {
+			if(grammarElements instanceof EObject) {
+				return new EObject[] {(EObject) grammarElements, grammarElement };
+			} else {
+				return ObjectArrays.concat((EObject[]) grammarElements, grammarElement);
+			}
+		}
+
+
+		@Override
+		public int hashCode() {
+			if (hashCode == -1)
+				hashCode = Objects.hash(grammarElement, grammarElements);
+			return hashCode;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if(o instanceof GrammarElementsInterner.InternKey) {
+				GrammarElementsInterner.InternKey interned = (GrammarElementsInterner.InternKey) o;
+				return Objects.equals(this.grammarElement, interned.grammarElement) 
+						&& Objects.equals(this.grammarElements, interned.grammarElements);
+			}
+			
+			return false;
+		}
+	}
+
+	Map<GrammarElementsInterner.InternKey, EObject[]> interningMap = Maps.newHashMap();
```

Comment:
This should probably be a concurrent map to support parallel parsing.

Reply:
Sorry, may fault: the interner is still used per resource.

The field should probably be protected and final, though.

Reply:
Fixed